modules/navier_stokes/src/fvkernels/PINSFVMomentumAdvection.C

```diff
@@ -0,0 +1,390 @@
+
+    ADRealVectorValue neighbor_velocity(_u_var->getNeighborValue(neighbor, *fi, elem_velocity(0)));
+    if (_v_var)
+      neighbor_velocity(1) = _v_var->getNeighborValue(neighbor, *fi, elem_velocity(1));
+    if (_w_var)
+      neighbor_velocity(2) = _w_var->getNeighborValue(neighbor, *fi, elem_velocity(2));
+
+    ADRealVectorValue interp_v;
+    Moose::FV::interpolate(Moose::FV::InterpMethod::Average,
+                           interp_v,
+                           elem_velocity,
+                           neighbor_velocity,
+                           *fi,
+                           elem_has_info);
+
+    // we are only interested in the interpolation coefficient for the Rhie-Chow element,
+    // so we just use the 'first' member of the returned pair
+    const auto advection_coeffs =
+        Moose::FV::interpCoeffs(_advected_interp_method, *fi, elem_has_info, interp_v);
+    ADReal temp_coeff = _rho * interp_v / eps_face * surface_vector * advection_coeffs.first;
+
+    // Now add the viscous flux. Note that this includes only the orthogonal component! See
+    // Moukalled equations 8.80, 8.78, and the orthogonal correction approach equation for
+    // E_f, equation 8.69
+    temp_coeff += mu / eps_face * surface_vector.norm() /
+                  (fi->neighborCentroid() - fi->elemCentroid()).norm();
+
+    // For internal faces the coefficient is the same for every velocity component.
+    for (const auto i : make_range(_dim))
+      coeff(i) += temp_coeff;
+  };
+
+  Moose::FV::loopOverElemFaceInfo(elem, _subproblem.mesh(), _subproblem, action_functor);
+
+  return coeff;
+}
+
+void
+PINSFVMomentumAdvection::interpolate(Moose::FV::InterpMethod m,
+                                     ADRealVectorValue & v,
+                                     const ADRealVectorValue & elem_v,
+                                     const ADRealVectorValue & neighbor_v)
+{
+  const Elem * const elem = &_face_info->elem();
+  const Elem * const neighbor = _face_info->neighborPtr();
+
+  if (onBoundary(*_face_info))
+  {
+#ifndef NDEBUG
+    bool flow_boundary_found = false;
+    for (const auto b_id : _face_info->boundaryIDs())
+      if (_flow_boundaries.find(b_id) != _flow_boundaries.end())
+      {
+        flow_boundary_found = true;
+        break;
+      }
+
+    mooseAssert(flow_boundary_found,
+                "INSFV*Advection flux kernel objects should only execute on flow boundaries.");
+#endif
+
+    v(0) = _u_var->getBoundaryFaceValue(*_face_info);
+    if (_v_var)
+      v(1) = _v_var->getBoundaryFaceValue(*_face_info);
+    if (_w_var)
+      v(2) = _w_var->getBoundaryFaceValue(*_face_info);
+
+    return;
+  }
+
+  Moose::FV::interpolate(
+      Moose::FV::InterpMethod::Average, v, elem_v, neighbor_v, *_face_info, true);
+
+  if (m == Moose::FV::InterpMethod::Average)
+    return;
+
+  // Avoid computing a pressure gradient near porosity jumps
+  if (!_smooth_porosity)
```

Comment:
I suppose this member just allows for some optimization to potentially avoid calling gradient reconstruction?

Reply:
it's not optimization, it's to zero the RC term when there s a porosity jump

Reply:
No I mean you would get the correct technical behavior without this user provided `_smooth_porosity` parameter right? If you removed `_smooth_porosity` then your direct checking of the porosity gradients would tell you whether you have porosity jumps or not. My concern is that relying on the user provided `smooth_porosity` value can be prone to error. For instance what if they say `smooth_porosity = true` when in fact it isn't? Then you you would go ahead and *not* zero the RC term near a jump.